include/dd/Package.hpp

```diff
@@ -1232,20 +1313,57 @@ namespace dd {
+                            if (!generateDensityMatrix || idx == 1) {
+                                // When generateDensityMatrix is false or I have the first edge I don't optimize anything and set generateDensityMatrix to false for all child edges
+                                m = multiply2(e1, e2, static_cast<Qubit>(var - 1), start, false);
+                            } else if (idx == 2) {
+                                // When I have the second edge and generateDensityMatrix == false, then edge[2] == edge[1]
+                                if (k == 0) {
+                                    if (edge[1].w.approximatelyZero()) {
+                                        edge[2] = ResultEdge::zero;
+                                    } else {
+                                        edge[2] = ResultEdge{edge[1].p, cn.getCached(edge[1].w.r->value, edge[1].w.i->value)};
+                                    }
+                                }
+                                continue;
+                            } else {
+                                m = multiply2(e1, e2, static_cast<Qubit>(var - 1), start, generateDensityMatrix);
+                            }
+                            if (k == 0 || edge[idx].w == Complex::zero) {
+                                edge[idx] = m;
+                            } else if (m.w != Complex::zero) {
+                                dEdge::applyDmChangesToEdges(&edge[idx], &m);
+                                auto old_e = edge[idx];
+                                edge[idx]  = add2(edge[idx], m);
+                                dEdge::revertDmChangesToEdges(&edge[idx], &e2);
+                                cn.returnToCache(old_e.w);
+                                cn.returnToCache(m.w);
+                            }
+                            //Undo modifications on density matrices
+                            dEdge::revertDmChangesToEdges(&e1, &e2);
+                        } else {
+                            auto m = multiply2(e1, e2, static_cast<Qubit>(var - 1), start);
+
+                            if (k == 0 || edge[idx].w == Complex::zero) {
+                                edge[idx] = m;
+                            } else if (m.w != Complex::zero) {
+                                auto old_e = edge[idx];
+                                edge[idx]  = add2(edge[idx], m);
+                                cn.returnToCache(old_e.w);
+                                cn.returnToCache(m.w);
+                            }
                         }
                     }
                 }
             }
-            e = makeDDNode(var, edge, true);
+            e = makeDDNode(var, edge, true, generateDensityMatrix);
+
+            //            if (r.p != nullptr && e.p != r.p) { // activate for debugging caching
+            //                std::cout << "Caching error detected in mul" << std::endl;
+            //            }
```

Comment:
superfluous comments

Reply:
Same as above: This statement is actually quite convenient to debug the operation cache, which is why I would like to keep it. However, I don't have a strong opinion on it.